nomad/state/state_store.go

```diff
@@ -4885,6 +4900,105 @@ func (s *StateStore) updateJobStabilityImpl(index uint64, namespace, jobID strin
 	return s.upsertJobImpl(index, nil, copy, true, txn)
 }
 
+func (s *StateStore) UpdateJobVersionTag(index uint64, namespace string, req *structs.JobApplyTagRequest) error {
+	jobID := req.JobID
+	jobVersion := req.Version
+	tag := req.Tag
+	name := req.Name
+
+	txn := s.db.WriteTxn(index)
+	defer txn.Abort()
+
+	// if no tag is present, this is a tag removal operation.
+	if tag == nil {
+		if err := s.unsetJobVersionTagImpl(index, namespace, jobID, name, txn); err != nil {
+			return err
+		}
+	} else {
+		if err := s.updateJobVersionTagImpl(index, namespace, jobID, jobVersion, tag, txn); err != nil {
+			return err
+		}
+	}
+
+	return txn.Commit()
+}
+
+func (s *StateStore) updateJobVersionTagImpl(index uint64, namespace, jobID string, jobVersion *uint64, tag *structs.JobTaggedVersion, txn *txn) error {
+	ws := memdb.NewWatchSet()
+
+	// Note: could use JobByIDAndVersion to get the specific version we want here,
+	// but then we'd have to make a second lookup to make sure we're not applying a duplicate tag name
+	versions, err := s.JobVersionsByID(ws, namespace, jobID)
+	if err != nil {
+		return err
+	}
+
+	duplicateVersionName := false
+	var job *structs.Job
+
+	for _, version := range versions {
+		// Allow for a tag to be updated (new description, for example) but otherwise don't allow a same-tagname to a different version.
+		if version.TaggedVersion != nil && version.TaggedVersion.Name == tag.Name && version.Version != *jobVersion {
+			duplicateVersionName = true
+			break
+		}
+		if version.Version == *jobVersion {
+			job = version
+		}
+	}
+
+	if duplicateVersionName {
+		return fmt.Errorf("tag %q already exists on a different version of job %q", tag.Name, jobID)
+	}
+
+	if job == nil {
+		return fmt.Errorf("job %q version %d not found", jobID, *jobVersion)
+	}
+
+	copy := job.Copy()
```

Comment:
`copy` is a Go built-in function, which you're shadowing here. That's not _wrong_ but we try to avoid it because it looks funny.

Reply:
I've got a [Copy() method on JobTaggedVersion](https://github.com/hashicorp/nomad/compare/main...epic/23794-golden-versions#diff-297649e31d8e88fd89ff63dac7aff450569cca19009616f08c1d73d1e8239725R4530-R4538), (which is the tag itself) but I think I don't understand how to use it here.

Or does this look funny because of the variable name `copy` itself?

Reply:
Update: TIL `shadowing`. Going to change the var name.